File: rust/src/events.rs
use crate::events::ffi_event::*;

#[cxx::bridge]
pub mod ffi_event {
    #[derive(Debug)]
    pub struct ModEvent {
        pub event_name: String,
        pub str_arg: String,
        pub num_arg: f64,
    }

    unsafe extern "C++" {
        include!("Events.h");
        fn AddTask_ModEvent(done: fn(ctx: ModEvent), ctx: ModEvent);
        unsafe fn SendEvent(event: ModEvent);
    }
}

impl ModEvent {
    pub fn new(event_name: &str, str_arg: &str, num_arg: f64) -> ModEvent {
        ModEvent {
            event_name: String::from(event_name),
            str_arg: String::from(str_arg),
            num_arg,
        }
    }
}

pub fn send_mod_event(event: ModEvent) {
    AddTask_ModEvent(
        |context| {
            unsafe {
                SendEvent(context);
            }
        },
        event,
    );
}

File: rust/src/Events.cpp
#pragma once
#include "Events.h"
#include "F4SE/API.h"
#include "lbug/src/logging.rs.h"
#include "RE/Bethesda/BSTSmartPointer.h"

// Outgoing Events

void AddTask_ModEvent(rust::Fn<void(ModEvent)> done, ModEvent ctx) noexcept {
    std::function<void()> task = std::move([done, ctx]() mutable {
        (*done)(ctx);
    });

    F4SE::GetTaskInterface()->AddTask(task);
}

void SendEvent(ModEvent event) noexcept {
    F4SE::GetPapyrusInterface()->GetExternalEventRegistrations(
        event.event_name.c_str(), 
        &event, 
        [](uint64_t handle, const char* scriptName, const char* callbackName, void* data
    ) {
		ModEvent* evt = static_cast<ModEvent*>( data );
		std::string strArg = (std::string) evt->str_arg;
		if (auto* vm = RE::GameVM::GetSingleton()) {
			vm->GetVM().get()->DispatchMethodCall(handle, scriptName, callbackName, nullptr, strArg, evt->num_arg);
		}
    });
}
